### PR TITLE
Engine interchangeable state storage

### PR DESCRIPTION
**Pull request by alexandraorth**
This pull request will allow you to use different `StateStorage` implementations in the cluster manager

+ `TaskStateZookeeperStorage` implements `StateStorage`
+ Renaming various `StateStorage` classes
+ Separated out the Zookeeper connection from the Zookeeper state storage class
+ Serializing the TaskState object to be stored in Zookeeper
+ Changed to using `mason` library in many Engine locations because it is serializable

### CODE REVIEW

**Comment by alexandraorth** [comment]
I'd just like to note that this is a very in-progress Engine. There are more refactors that i'm working on right now which will streamline how things are instantiated and called yet again.

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/TaskStateStorage.java [hunk 1]
```diff
@@ -19,52 +19,25 @@
 package ai.grakn.engine.backgroundtasks;
 
 import javafx.util.Pair;
-import org.json.JSONObject;
-
-import java.time.Instant;
 import java.util.Set;
 
-public interface StateStorage {
+public interface TaskStateStorage {
```
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/TaskStateStorage.java]: Just noticed that this is missing a class header for java docs.

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/distributed/ClusterManager.java [hunk 2]
```diff
@@ -70,6 +72,10 @@ public ClusterManager() {
             LOG.debug("Starting Cluster manager on " + ENGINE_ID);
 
             startZookeeperConnection();
+
+            //TODO start state storage, decide using reflection
```
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/distributed/ClusterManager.java]: Are we still going to be maintaining the in memory and grakn state storage? Do we still need both?

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/standalone/StandaloneTaskManager.java [hunk 3]
```diff
@@ -191,24 +195,27 @@ public TaskManager stopTask(String id, String requesterName) {
...
         return () -> {
             try {
-                task.start(saveCheckpoint(id), stateStorage.getState(id).configuration());
+                task.start(saveCheckpoint(state), state.configuration());
 
                 stateUpdateLock.lock();
-                if(stateStorage.getState(id).status() == RUNNING) {
-                    stateStorage.updateState(id, COMPLETED, EXCEPTION_CATCHER_NAME, null, null, null, null);
+                if(state.status() == RUNNING) {
+                    stateStorage.updateState(state.status(COMPLETED).statusChangedBy(EXCEPTION_CATCHER_NAME));
```
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/standalone/StandaloneTaskManager.java]: Look at all those nulls die :)

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstatestorage/TaskStateGraphStore.java [hunk 4]
```diff
@@ -178,38 +164,27 @@ public Boolean updateState(String id, TaskStatus status, String statusChangeBy,
     }
 
     public TaskState getState(String id) {
-        if(id == null) {
-            return null;
-        }
-
         Optional<TaskState> result = attemptCommitToSystemGraph((graph) -> {
-            Instance instance = graph.getConcept(ConceptId.of(id));
+            Instance instance = graph.getResourcesByValue(id).iterator().next().owner();
```
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstatestorage/TaskStateGraphStore.java]: Do you know the resource type? you could make this MUCH faster and scale better by saying `graph.getResourceType("My Type").getResource("Value")`

File: grakn-test/src/test/java/ai/grakn/test/engine/backgroundtasks/TaskStateGraphStoreTest.java [hunk 5]
```diff
@@ -74,78 +76,53 @@ public void testStoreRetrieve() {
...
-
-        id = stateStorage.newState(TestTask.class.getName(), this.getClass().getName(), null, true, 10000, new JSONObject());
-        assertNull(id);
-
-        id = stateStorage.newState(TestTask.class.getName(), this.getClass().getName(), Instant.now(), null, 10000, new JSONObject());
-        assertNull(id);
-    }
-
-    @Test
     public void testUpdate() {
```
Comment (fppt) [inline grakn-test/src/test/java/ai/grakn/test/engine/backgroundtasks/TaskStateGraphStoreTest.java]: In an effort to make tests more clear. Test update what? What in the task are you checking is being updated?

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/distributed/ClusterManager.java (hunk 2 above)
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/distributed/ClusterManager.java]: We are going to be maintaining both. The InMemory will be used for testing non-engine tests that require engine to be running (GraqlShell, Migration and Analytics), which should speed tests up significantly. Boris also suggested that our single-machine dist could use it.

File: grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstatestorage/TaskStateGraphStore.java (hunk 4 above)
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/backgroundtasks/taskstatestorage/TaskStateGraphStore.java]: Oooh, I do! I'll make this change :)

File: grakn-test/src/test/java/ai/grakn/test/engine/backgroundtasks/TaskStateGraphStoreTest.java (hunk 5 above)
Comment (alexandraorth) [inline grakn-test/src/test/java/ai/grakn/test/engine/backgroundtasks/TaskStateGraphStoreTest.java]: Good idea. Have done this in my current code- will be in next pull.

**Comment by fppt** [comment]
@alexandraorth rebase please, also if you could do the resource type optimisation I mentioned above.

**Pull request closed without merging**